From a6f33c7b88c8cd482c0b006917e66507f4bc8f56 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:45 -0400
Subject: [powerpc] Check cpu_thread_in_subcore() in __cpu_up()

Message-id: <1410545655-205645-537-git-send-email-dzickus@redhat.com>
Patchwork-id: 94386
O-Subject: [RHEL7 PATCH 536/626] powerpc: Check cpu_thread_in_subcore() in __cpu_up()
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 6f5e40a3001d2497a134386a173e3ec3fdf2ad0b
Author: Michael Ellerman <mpe@ellerman.id.au>
Date:   Fri May 23 18:15:28 2014 +1000

    powerpc: Check cpu_thread_in_subcore() in __cpu_up()

    To support split core we need to change the check in __cpu_up() that
    determines if a cpu is allowed to come online.

    Currently we refuse to online cpus which are not the primary thread
    within their core.

    On POWER8 with split core support this check needs to instead refuse to
    online cpus which are not the primary thread within their *sub* core.

    On POWER7 and other systems that do not support split core,
    threads_per_subcore == threads_per_core and so the check is equivalent.

    Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>
    Signed-off-by: Michael Neuling <mikey@neuling.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/smp.c b/arch/powerpc/kernel/smp.c
index 5cdd9eb..6af946e 100644
--- a/arch/powerpc/kernel/smp.c
+++ b/arch/powerpc/kernel/smp.c
@@ -490,7 +490,7 @@ int __cpu_up(unsigned int cpu, struct task_struct *tidle)
   * Don't allow secondary threads to come online if inhibited
   */
  if (threads_per_core > 1 && secondaries_inhibited() &&
-     cpu % threads_per_core != 0)
+     cpu_thread_in_subcore(cpu))
   return -EBUSY;
 
  if (smp_ops == NULL ||
-- 
1.7.1